### ISSUE TITLE
Proposal: New lint and `package:meta` annotation to default to tight class capabilities

### ISSUE THREAD

**Issue by parlough** [dart-lang/sdk#54656]
I think for some library or SDK developers, it makes sense to be explicit about how you open the capabilities of your public classes, so you are forced to consider how support will evolve. However, Dart defaults to a permissive model and doesn't have an `open` keyword.

To help library developers make conscious decisions early on, a lint could be introduced to flag all public class declarations without a capability class modifier (`final`, `base`, `interface`, or `sealed`). The author of the class would then need to consider their intentions with the class and either add one of those keywords or add an `@open` annotation.

I think under this rule, the following class declarations, the following class declarations would trigger the lint:

```
class A {}
abstract class B {}
mixin class C {}
abstract mixin class D {}
mixin {}
```

**Comment by parlough** [comment]
I'm happy to implement this, but I'd like to verify others besides me see it as worthwhile.

\cc @lrhn @natebosch @davidmorgan Do you have any thoughts based on your experience maintaining Dart libraries over the years? I know it might not be particularly useful for you, but I'm curious about your thoughts in terms of benefiting package developers across the Dart ecosystem. Maybe even those that aren't yet super familiar with class modifiers.

One potential concern often raised with functionality like this is developers restricting capabilities unnecessarily, frustrating downstream consumers.

**Comment by lrhn** [comment]
If you want the lint for yourself, go for it,  but I would argue hard against adding it to the recommended lints.

The general consensus when we (IIRC it was @munificent) did a survey, was that people wanted permissive as the default.

Whether that's right for a project probably depends on the kind of project.

An app shouldn't need to worry about other people extending their stuff, because they don't need to release the source at all. Open is perfect.

A helper library package may or may not want to close things down, depending on their future intents, and what kind of package it is.
A complicated framework is likely different from just supplying a few dyrene-alone data structures.

So this lint would not be recommended.
(Which implies it won't be in the core lints either, since the recommended lints extend those.)

A lint which enforces a specific strategy, which a user can choose or not to apply to their project, is what lints are for.
The recommended lints are just overs that we believe are generally applicable and valuable. Some of those should probably just have been plain analyzer warnings.

What the lint would do is to warm if any `class`, `mixin` or `mixin class` declaration does not have a `final`, `base`, `interface` or `sealed` modifier, and also does not have an `@open` (or `@reopen`?) annotation.
(Or if it has both a modifier and an `@open` annotation.)

Basically it makes `@open` behave like a fifth modifier, and requires every declaration to have one modifier.
(If that's the way it's presented, then `@reopen` should probably not count, and opening a subclass of an `interface` class would require `@reopen @open`, if you have the `annotate_reopen` lint enabled as well.)

SGTM, but not a lint I'd probably use myself.

**Comment by munificent** [comment]
> The general consensus when we (IIRC it was @munificent) did a survey, was that people wanted permissive as the default.

Yeah, it was a surprisingly strong consensus too. Even among package authors whose APIs are widely consumed and who likely want to be more restrictive, the overall preference was to default to permissiveness. I guess it's just an appreciated part of Dart's vibe.

> I wonder if @munificent would consider a style rule along the lines of "if a class in your public API is designed for extension, document it"; and then maybe that leads into this lint, even if it's not quite enough to have it be enabled by default.

Do you mean [this rule](https://dart.dev/effective-dart/design#avoid-extending-a-class-that-isnt-intended-to-be-subclassed)? :)

I'm not opposed to this lint in principle. (For many years, I personally felt that the default should be closed, but I was persuaded by our community to change my opinion.) I do worry a little that `open` is too short and common of an identifier to stuff into package:meta. Maybe `openType`? Or maybe just have users who want this lint enabled use `// ignore` to ignore it on the classes they do intend to be open?

**Comment by davidmorgan** [comment]
Thanks Bob! Almost that rule, it turns out you already have exactly what I was going for in the [next one](https://dart.dev/effective-dart/design#do-document-if-your-class-supports-being-extended) ;) "DO document if your class supports being extended". 

It's even "DO" rather than "PREFER", which usually makes a nice candidate for a lint. From the wording of the rule, a lint based on it would allow an extensible class only if it has "Base" in the name of if it mentions extending in the docs.

I kinda like the idea. (Combined with "and don't fire under src/" mentioned above).

**Comment by munificent** [comment]
> From the wording of the rule, a lint based on it would allow an extensible class only if it has "Base" in the name

Yeah, that's an easy one.

> of if it mentions extending in the docs.

Detecting this correctly could be pretty hard.

**Comment by davidmorgan** [comment]
> > of if it mentions extending in the docs.
> 
> Detecting this correctly could be pretty hard.

Eh, I would just match on "extend(ing)? this class" and ship it, people can send a PR adding a more forms if they want ;)

**Comment by lrhn** [comment]
> only if it has "Base" in the name of if it mentions extending in the docs.

I'm generally moving away from using `Base` in names, because today it is possible to make something like `Iterable` which is all of an interface, a base/skeleton class and a mixin.
It's open because it intends all the possible uses. (Even has a const constructor.) So makes ending in `Base` is probably not as important a signal as it has been.

A lint for this rule would be that _if_ a class can be extended (public generative constructor, marked `base` or "open"), and is not marked `base` (so it's open), then it should document that it's intended for extending.
Either by annotating with `@open`, or whatever this issue ends up with, or by having any of a known set of naming styles or documentation phrases. (Which can also be on constructor.)

It is doable. I think I'll even be in favor of recommending it, if we get the heuristics right

This part of the discussion is focusing on extending, which is the most common problem in other languages, but only because they have separate interface declarations. Allowing implementation without intending it is often a worse accident for a package author, because it restricts their ability to evolve the class.
Adding a method to a base class _can be_ breaking. Adding one to an interface that is implemented _is_ breaking.

So I'm more worried about an open class being implementable, and it's harder to document that implementation is allowed. Mainly because being implementable is so fundamental in Dart that we don't have any standard phrases for saying that it is.
And then I'd want lint users to add `@open` on every open class anyway, which would silence the above lint too, making it effectively subsumed by an "explicitly mark open classes" lint.
